SMC Capability

Furao opened the pull request:
Add a new capability which certain threads can invoke so seL4 can make SMC calls on ARM platforms in EL2 (virtualized mode) on the thread’s behalf.

See https://sel4.atlassian.net/browse/RFC-9

yyshen commented on libsel4/sel4_arch_include/aarch64/interfaces/sel4arch.xml:
```diff
@@ -44,6 +44,16 @@
+        <member name="x4"/>
+        <member name="x5"/>
+        <member name="x6"/>
+        <member name="x7"/>
```
might need to state that this only works for SMC calls that follow the SMC calling convention.

yyshen commented on src/arch/arm/config.cmake:
```diff
@@ -194,6 +194,12 @@ config_option(
 
+config_option(
+    KernelAllowSMCCalls ALLOW_SMC_CALLS "Allow virtualized guests to make SMC calls"
+    DEFAULT OFF
```
probably need to enable the option that traps SMC calls?

yyshen commented on src/arch/arm/object/smc.c:
```diff
@@ -0,0 +1,93 @@
+    switch (label) {
+    case ARMSMCCall:
+        badge = cap_smc_cap_get_capSMCBadge(cap);
+        smc_func_id = getSyscallArg(0, buffer);
```
so the badge is used to represent the authorised smc_func_id for a minted SMC Cap. by default, the original  SMC Cap can authorise all smc func calls, right?

lsf37 replied:
According to the code, badge 0 authorises all, which would be the same convention as for endpoints. I think that makes sense. We should check though, if smc call 0 is reserved or may be needed, because otherwise anyone who needs to do that call can do all calls.

yyshen commented on src/arch/arm/object/smc.c:
```diff
@@ -0,0 +1,93 @@
+    register seL4_Word r5 asm("x5") = arg[5];
+    register seL4_Word r6 asm("x6") = arg[6];
+    register seL4_Word r7 asm("x7") = arg[7];
+    asm volatile("smc #0\n"
```
have you considered to release the lock for the SMP kernel before the call and acquire the lock again after the call. The smc call can run arbitrary long.

lsf37 replied:
Hm, that is an important observation -- I hadn't thought about WCET on these yet at all, and without knowing more about individual calls, the kernel WCET for a system with SMC enabled is now arbitrarily high. That is even true on uni-core, but it is worse on SMP, because the kernel is blocked for all cores during that time.

This is different to other user-level behaviour that can take arbitrarily long, because this is now kernel time.

I don't think that releasing the lock while inside the kernel works in the current design: it wouldn't be a big-lock any more, and when you re-acquire the lock you would have not guarantee that the current TSC, SC, or IPC buffer still exists (some other core could have retyped them, for instance). I'm sure that could be worked-around, but it would definitely break the big-lock idea. And even if that should somehow work, it would still not solve the uni-core case.

That might throw a real spanner in the works. Do we need to turn this into some kind of more asynchronous mechanism?

axel-h replied:
if the short term goal is, that the feature is to be merged (behind a config option) , I'm fine with the comment https://sel4.atlassian.net/browse/RFC-9?focusedCommentId=19762 from @lsf37 that for now the system designer needs to be aware of the potential pitfalls. We can look into this as a follow-up then.

Furao replied:
The warning message has been added here: https://github.com/seL4/seL4/pull/701/files#diff-bc9402a3a9f4e827bf5919980dd881d416c5a6225a82a380af72117a605ae299R198-R203

yyshen commented on libsel4/include/sel4/bootinfo_types.h:
```diff
@@ -24,11 +24,12 @@ enum {
     seL4_CapDomain              = 11, /* global domain controller cap */
     seL4_CapSMMUSIDControl      = 12,  /*global SMMU SID controller cap, null cap if not supported*/
     seL4_CapSMMUCBControl       = 13,  /*global SMMU CB controller cap, null cap if not supported*/
+    seL4_CapSMC                 = 14,  /*global SMC cap, null cap if not supported*/
```
It looks like one SMC cap is used for all CPU cores, not sure if this is coarse-grained.

lsf37 replied:
That looks fine to me, as long as the intended pattern is that the system initialiser makes badged copies of this cap, handing these out to whoever needs them, restricted to core, sub-system or whichever is needed. Usually the 0-badge version of the cap should stay in the initialiser, or in a high-privilege management component.

That scenario does suggest that we might want more derivation levels for this cap, though, like for endpoint caps.

axel-h commented on libsel4/arch_include/arm/interfaces/sel4arch.xml:
```diff
@@ -23,7 +23,7 @@
             <param dir="in" name="attr" type="seL4_ARM_VMAttributes">
-                <description> 
+                <description>
                     VM Attributes for the mapping. <docref>Possible values for this type are given in <autoref label="ch:vspace"/>  .</docref>
```
Could you move these space changes to a dedicated commit?

Furao replied:
Yeah, I can do that.

axel-h commented on libsel4/include/sel4/bootinfo_types.h:
```diff
@@ -24,11 +24,12 @@ enum {
     seL4_CapDomain              = 11, /* global domain controller cap */
     seL4_CapSMMUSIDControl      = 12,  /*global SMMU SID controller cap, null cap if not supported*/
     seL4_CapSMMUCBControl       = 13,  /*global SMMU CB controller cap, null cap if not supported*/
+    seL4_CapSMC                 = 14,  /*global SMC cap, null cap if not supported*/
```
Is this a potentially breaking change that should be noted in the readme?

Furao replied:
That is a good point I hadn't considered. Adding to the end avoids breaking old layouts, but that becomes much harder to do with the sched_context cap getting configured out. Maybe positions should be constant whether their features are enabled or not? 
ex:
```
enum {
    seL4_CapNull                =  0, /* null cap */
    seL4_CapInitThreadTCB       =  1, /* initial thread's TCB cap */
    seL4_CapInitThreadCNode     =  2, /* initial thread's root CNode cap */
    seL4_CapInitThreadVSpace    =  3, /* initial thread's VSpace cap */
    seL4_CapIRQControl          =  4, /* global IRQ controller cap */
    seL4_CapASIDControl         =  5, /* global ASID controller cap */
    seL4_CapInitThreadASIDPool  =  6, /* initial thread's ASID pool cap */
    seL4_CapIOPortControl       =  7, /* global IO port control cap (null cap if not supported) */
    seL4_CapIOSpace             =  8, /* global IO space cap (null cap if no IOMMU support) */
    seL4_CapBootInfoFrame       =  9, /* bootinfo frame cap */
    seL4_CapInitThreadIPCBuffer = 10, /* initial thread's IPC buffer frame cap */
    seL4_CapDomain              = 11, /* global domain controller cap */
    seL4_CapSMMUSIDControl      = 12,  /*global SMMU SID controller cap, null cap if not supported*/
    seL4_CapSMMUCBControl       = 13,  /*global SMMU CB controller cap, null cap if not supported*/
    seL4_CapInitThreadSC        = 14, /* initial thread's scheduling context cap */
    seL4_CapSMC                    = 15,  /*global SMC cap, null cap if not supported*/
    seL4_NumInitialCaps         = 16
};
```

Then if they feature isn't enabled, it just shows up as a nullcap?

axel-h replied:
@lsf37, @kent-mcleod; What do you think about this? 

A change here might also require updating `SIMPLE_SKIPPED_INIT_CAPS` here:
- https://github.com/seL4/seL4_libs/blob/master/libsel4simple/arch_include/arm/simple/arch/simple.h#L17-L19 (note that there is no RISC-V here, it uses the ARM files)
- https://github.com/seL4/seL4_libs/blob/master/libsel4simple/arch_include/x86/simple/arch/simple.h#L17-L23

Furao replied:
@lsf37, @kent-mcleod; ping.

Furao replied:
@lsf37, @kent-mcleod; pinging again.

lsf37 replied:
Apologies for the long silence. As commented in #1088, from the verification side that would be Ok.

kent-mcleod replied:
Once this is rebased on the changes from #1088, this should be resolved?

axel-h commented on src/arch/arm/64/object/objecttype.c:
```diff
@@ -495,6 +501,10 @@ exception_t Arch_decodeInvocation(word_t label, word_t length, cptr_t cptr,
     case cap_cb_cap:
         return decodeARMCBInvocation(label, length, cptr, slot, cap, call, buffer);
 #endif /*CONFIG_ARM_SMMU*/
+#ifdef CONFIG_ALLOW_SMC_CALLS
```
I think we have to add this in line 488 also

Furao replied:
CONFIG_ALLOW_SMC_CALLS has a dependency on CONFIG_ARM_HYPERVISOR_SUPPORT, so it works how it is. However, we can be more explicit and add it into 488 if you would like.

axel-h replied:
I think having this explicitly improves maintainability. How strong us the dependency of the SMC feature on the Hypervisor feature - could they be separated one day?

Furao replied:
We have actually already seen a use case to uncouple this from virtualization support, so we updated the specified ifdef line and changed the config dependency.

Furao commented:
@axel-h Ping

axel-h commented:
Could you rebase this in the current master?

Furao commented:
Ok, rebased.

Furao commented on src/arch/arm/config.cmake:
```diff
@@ -194,6 +194,17 @@ config_option(
+    since the supervisor is a higher privilege level and can make any change \
+    in the system, the proofs can no longer be guaranteed."
+    DEFAULT OFF
+    DEPENDS "KernelSel4ArchAarch64"
```
Yeah, that sounds like a good idea. I will make that update.

Furao replied:
Updated this dependency in the latest.

kent-mcleod commented:
@Furao, I added some suggested changes but can't PR off this PR due to a rebase.  https://github.com/seL4/seL4/compare/master...kent-mcleod:seL4:kent/smc.  Are you happy to pull these additional commits in?  In addition to some changes to the invocation implementation to more closely follow the structures required by the verification specs, I've made it so that seL4_CNode_Revoke should do the right thing:
- Revoking the root's original smc cap should delete all SMC caps in the system,
- Revoking the first badged smc cap should delete any SMC caps derived from it.

kent-mcleod commented on src/object/cnode.c:
```diff
@@ -806,6 +806,21 @@ bool_t PURE isMDBParentOf(cte_t *cte_a, cte_t *cte_b)
         break;
     }
 
+#ifdef CONFIG_ALLOW_SMC_CALLS
```
This would benefit from some arch-specific functions to call out to, but they don't currently exist... Thoughts?

Furao replied:
I suppose it could also be wrapped in an architecture ifdef so that if we ever make an architecture specific `cnode.c` file, then it is obvious that chunk of code should move.

Indanz commented on src/object/cnode.c:
```diff
@@ -806,6 +806,21 @@ bool_t PURE isMDBParentOf(cte_t *cte_a, cte_t *cte_b)
+        if (badge == 0) {
+            return true;
+        }
+        return
```
The formatting of this return is very weird. I know that cap_notification_cap does the same, but I don't think there is a reason to spread this further.

Furao replied:
Do you just mean that its weird that the keyword is on its own line?

Indanz replied:
Yes.

kent-mcleod replied:
I thought that the original weirdness is due to an autostyle edge case. Turns out it's not, as the autostyle accepts it on the same line.. Agree we should change it.

Indanz commented on libsel4/sel4_arch_include/aarch64/interfaces/sel4arch.xml:
```diff
@@ -207,4 +217,18 @@
+                Tell the microkernel to make the real SMC call.
+            </brief>
+            <description>
+                <docref>See <autoref label="sec:read_write_registers"/></docref>
```
Is this reference still up-to-date, considering that `seL4_ARM_SMCContext` has been added?

Furao replied:
Updated the description accordingly.

Indanz commented on src/arch/arm/object/smc.c:
```diff
@@ -0,0 +1,89 @@
+    register seL4_Word r6 asm("x6") = arg[6];
+    register seL4_Word r7 asm("x7") = arg[7];
+    asm volatile("smc #0\n"
+                 : "+r"(r0), "+r"(r1), "+r"(r2), "+r"(r3)
+                 : "r"(r4), "r"(r5), "r"(r6), "r"(r7)
```
According to DEN0028E_SMC_Calling_Convention_1.4eac0:

- A Function Identifier is passed in register W0.
- Arguments are passed in registers W1-W7.
- Results are returned in W0-W7.
- Registers W4-W7 must be preserved unless they contain results, as specified in the function definition.

(Same for 64-bit calls, but with X0-X17 instead.)

So I think both the output and input lists should be extended to the r0-r7 registers, both with `+r`.

Also, if only the 32-bit calling convention is supported, the code should check that bit 30 is not set in `smc_func_id`.

Furao replied:
So we are actually targeting a specific version of the SMCCC spec. v1.0. This document can be viewed here: https://developer.arm.com/documentation/den0028/a/?lang=en (The `a` in the URL should give you the correct version when you download.). You can query the SMCCC version number on your platform with an SMC call. We did our development on a Zynq UltraScale+ MPSoC, which implements spec v1.0. 

In this version, r4-r7 is just scratch space, so it won't ever be a result. 
Also, since we are using `x0-x7`, we are actually using the 64-bit calling convention.

This does mean that if future processors conform to newer calling convention versions, than we need to create version  specific implementations.

Indanz replied:
We don't want version specific implementations. This is a new multi-platform feature, it should be as future proof as possible, which means following the latest spec. If in the future a platform does return data in registers x4-x7, with the current code it will silently clobber a register and that will be extremely hard to debug if it causes weird problems.

If you do follow the 64-bit calling convention, a lot more registers get clobbered, see table 3-1. The normal procedure call conventions are used, registers `x0-x17`  may be modified. On the upside, you don't have to check bit 30, as you can do 32-bit calls with the 64-bit convention.

With that in mind, maybe it's better to add a `do_smc` function to assembly and call it from here, I think that will get all clobbers and assumptions right automatically.

Indanz replied:
I'm fine with limiting it to `x0-x7` for now, as it will take ages before more is practically needed, and when the limit is reached, it's easy to extend. On the other hand, it's probably not much more work to extend it now.

>  I am not entirely sure what an assembly function gets us over the inline assembly we are already doing. The context still needs to be transferred in/out of the registers?

Doing the SMC call would be just a normal function call then. But I forgot that there is a need to transfer the result too, so the advantage would be much less, if any.

Furao replied:
Playing around with with using all 18 registers revealed to me that inline assembly calls can only have 30 operands, so 18 inputs + 18 outputs doesn't work. Path of least resistance is probably to just modify `r4-r7` to be `r+`, so it is an input and output. 

Or if we really wanted to support all 18 registers, we would need to write a proper assembly function. Thoughts?

Furao replied:
I implemented the simpler approach with my latest push.

Indanz replied:
Almost there, you also need to add all other modified registers to the clobber list, see Table 3-1.

Furao replied:
I appreciate the input. This implementation is much more future-proof now.

Furao commented on libsel4/sel4_arch_include/aarch64/interfaces/sel4arch.xml:
```diff
@@ -207,4 +217,18 @@
             </error>
         </method>
     </interface>
+    <interface name="seL4_ARM_SMC" manual_name="SMC" cap_description="Capability to allow threads to interact with SMC calls.">
```
Agreed. Updated.

kent-mcleod commented:
> Yeah, I can pull those in.

Thanks.  I'd be happy to apply any further fixes myself, but the PR branch permissions don't allow modifications by reviewers.

Indanz commented on src/arch/arm/object/smc.c:
```diff
@@ -0,0 +1,90 @@
+    asm volatile("smc #0\n"
+                 : "+r"(r0), "+r"(r1), "+r"(r2), "+r"(r3),
+                 "+r"(r4), "+r"(r5), "+r"(r6), "+r"(r7)
+                 :
+                 : "memory");
```
```suggestion
                 :: "x8", "x9", "x10", "x11", "x12", "x13", "x14", "x15", "x16", "x17", "memory");
```

Furao replied:
Made the suggested change

axel-h commented:
I might be good to mention the new `seL4_CapSMC` and [RFC-9](https://sel4.atlassian.net/browse/RFC-9) in the [CHANGES](https://github.com/seL4/seL4/blob/master/CHANGES) file also.

axel-h commented on src/arch/arm/config.cmake:
```diff
@@ -208,6 +208,17 @@ config_option(
+    KernelAllowSMCCalls ALLOW_SMC_CALLS "Allow components to make SMC calls.
+    WARNING: Some platforms require SMC calls to actually be made to boot VMs \
+    correctly, however allowing SMC calls causes a couple of issues. Since seL4 \
+    cannot pre-empt the supervisor the WCET is no longer guaranteed. Also, \
```
```suggestion
    cannot pre-empt the supervisor, the WCET is no longer guaranteed. Also, \
```

axel-h replied:
Also, I think we should use "monitor" (or "secure monitor") instead of "supervisor" here.

Furao replied:
Updated both suggestions.

Furao commented on src/arch/arm/config.cmake:
```diff
@@ -208,6 +208,17 @@ config_option(
 
+config_option(
+    KernelAllowSMCCalls ALLOW_SMC_CALLS "Allow components to make SMC calls.
+    WARNING: Some platforms require SMC calls to actually be made to boot VMs \
```
Agreed. I removed that line.

Indanz replied:
Agreed. Other than doing standard PSCI calls, SCM is used for platform specific things. People needing SCM will know whether they need it or not.

Furao commented:
> I might be good to mention the new `seL4_CapSMC` and [RFC-9](https://sel4.atlassian.net/browse/RFC-9) in the [CHANGES](https://github.com/seL4/seL4/blob/master/CHANGES) file also.

Added a new commit to update CHANGES file.

Furao commented:
I just rebased onto the latest and I believe I have addressed all of the open comments. Would be nice to get this through.

axel-h reviewed (approved):
From my side this is fine now. I lack the resources to test this in more detail, tough. 
You could squash the last commit with the CHANGES update into the commit that adds the capability.

kent-mcleod commented:
> I just rebased onto the latest and I believe I have addressed all of the open comments. Would be nice to get this through.

I'll wait a couple more days for feedback from other reviewers and otherwise merge.

Ivan-Velickovic reviewed (commented):
The single comment I had has been addressed, but I’m not qualified to review/approve the actual contents of the patch so I’ll leave that for the others.

lsf37 reviewed (approved):
No further issues from my side. If we had a model of what SMC calls do this does even look fairly verifiable. 

(The preprocess check triggers on the change of `seL4_NumInitialCaps` which should not affect the proofs)

lsf37 commented:
Have remove the proof-test label to preserve AWS instances, because the proofs have passed with this PR now. It looks like the build failed with sel4test because the PR needs to be rebased first to work with the latest libsel4.

Indanz commented on src/arch/arm/object/smc.c:
```diff
@@ -8,18 +8,17 @@
         arg[i] = getSyscallArg(i, buffer);
     }
 
-    ipcBuffer = lookupIPCBuffer(true, NODE_STATE(ksCurThread));
```
I don't think you can safely remove this, however tempting:

 > Require the input length to be >= NUM_SMC_REGS. This ensures that any  accesses to the IPC buffer memory is safe (if the IPC buffer isn't  present then the message length will be truncated to n_msgRegisters).

This misses the corner case where the IPC buffer is mapped read-only.

kent-mcleod replied:
Yea you're right good catch. TIL that IPC buffer caps can be without write rights.  (seems like such an anti-feature)

Furao replied:
Added it back in

Indanz reviewed (commented):
I'd add a back reference to this pull request to the first commit, so people can find this easily when going through the code or git log.

Furao commented on libsel4/sel4_arch_include/aarch64/interfaces/sel4arch.xml:
```diff
@@ -207,4 +217,20 @@
+                struct. The microkernel makes the SMC call and then returns the results as a
+                new seL4_ARM_SMCContext.
+            </description>
+            <param dir="in" name="smc_args" type="seL4_ARM_SMCContext"
+                description="The structure that has the provided arguments."/>
+            <param dir="out" name="smc_response" type="seL4_ARM_SMCContext"
+                description="The structure to capture the responses."/>
```
Correct, `inout` is not supported. 

`syscall_stub_gen.py` does the parsing:
```py
assert (param_dir == "in") or (param_dir == "out")
```

Indanz replied:
Oh well, so be it, it's not the end of the world and can be changed later if we want.

Furao commented:
> From my side this is fine now. I lack the resources to test this in more detail, tough. You could squash the last commit with the CHANGES update into the commit that adds the capability.

Squashed that change with the first one.

> I'd add a back reference to this pull request to the first commit, so people can find this easily when going through the code or git log.

Updated the first commit message with the link to this PR.

Indanz commented on src/arch/arm/object/smc.c:
```diff
@@ -0,0 +1,94 @@
+    }
+
+    setThreadState(NODE_STATE(ksCurThread), ThreadState_Restart);
+    return invokeSMCCall(buffer, call);
```
`buffer` is not used any more, perhaps move the `lookupIPCbuffer()` call to here and pass that instead of buffer? Or remove the buffer argument, like you had before.

Indanz reviewed (approved):
Looks good now.

Would be nice if:

https://github.com/seL4/sel4test/pull/85
https://github.com/seL4/sel4test/pull/101

would get merged together, but it's fine if they get merged upstream separately. Could you do a review of Kent's SMC tests?

axel-h commented on src/arch/arm/object/smc.c:
```diff
@@ -0,0 +1,94 @@
+
+    ipcBuffer = lookupIPCBuffer(true, NODE_STATE(ksCurThread));
+
+    register seL4_Word r0 asm("x0") = arg[0];
```
Maybe this should have a compile time assert that `NUM_SMC_REGS` is at least 8. It's unlikey to ever change, but may catch cases where somebody is experimenting. I'd expect the compiler to issue a out-of-warning in this case automatically, though - but does this also break the build with the current defaults?

kent-mcleod replied:
If you want to make that in a separate PR then I'd be happy to accept it.

axel-h replied:
will do

Indanz commented on include/arch/arm/arch/object/structures.h:
```diff
@@ -12,6 +12,16 @@
+    case cap_smc_cap:
+        return (cap_smc_cap_get_capSMCBadge(derivedCap) !=
+                cap_smc_cap_get_capSMCBadge(srcCap));
+
```
I would remove this empty line, it looks a bit weird.

Ivan-Velickovic commented:
My understanding is that this PR does not support the SMC64 calling convention, is that correct?

Ivan-Velickovic commented:
> > My understanding is that this PR does not support the SMC64 calling convention, is that correct?
> 
> No, it's not. You can do both 32-bit and 64-bit SMC calls with this interface. However, it's only implemented for Aarch64, not 32-bit ARM.

The reason I asked is being the SMC specification I was looking at (downloaded from here https://developer.arm.com/documentation/den0028/f?lang=en) says that for SMC64/HVC64 (section 2.7), registers x1-x17 are used for arguments and x0-17 are used for results. This patch takes in x0-x7 as arguments/results.

I *think* I've found a case where I need SMC64, so if I'm happy to add support for the SMC64 calling convention, assuming I've understood everything correctly.

Ivan-Velickovic commented:
> Technically, we are supporting 32-bit and 64-bit SMC calls but for an earlier version of the spec (v1.0) which didn't include x8-x17 for arguments or results. I looked at adding the rest of the registers, but ran into an issue with inline assembly.

Yep, I just read through the spec changes and saw that in August 2020 it was introduced to use x8-x17 as arguments/results. That makes sense, now. Although, we should document this. I assume then issue B/version 1.1 of the spec was used?

> To support this, you need to do a refactor and write a new function as just assembly.

Yes I ran into the same issue.

Ivan-Velickovic commented:
Ah GitHub hides all the existing conversations on the calling convention which is why I missed it, my bad.

Furao commented:
> Ah GitHub hides all the existing conversations on the calling convention which is why I missed it, my bad.

No problem. I knew to look for the comment and it was hard for me to find as well.

> > Technically, we are supporting 32-bit and 64-bit SMC calls but for an earlier version of the spec (v1.0) which didn't include x8-x17 for arguments or results. I looked at adding the rest of the registers, but ran into an issue with inline assembly.
> 
> Yep, I just read through the spec changes and saw that in August 2020 it was introduced to use x8-x17 as arguments/results. That makes sense, now. Although, we should document this. I assume then issue B/version 1.1 of the spec was used?

Yes documentation on the version would be good. Yes, I think 1.1 is technically the latest one we support.
